File: ios/LynxShellKit/Telemetry/TelemetryModels.swift
import Foundation

/// Wire Schema 3.0.0 中允许的 Bundle 来源。
public enum LynxTelemetryBundleSource: String, Codable, Equatable {
    case ota
    case directHttps = "direct_https"
    case assets
    case localFile = "local_file"
}

/// 允许业务属性进入事件的四种基础类型；Native 字段不能由页面覆盖。
public enum LynxTelemetryValue: Codable, Equatable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case null

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let bool = try? container.decode(Bool.self) {
            self = .bool(bool)
        } else if let number = try? container.decode(Double.self) {
            self = .number(number)
        } else {
            self = .string(try container.decode(String.self))
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case let .string(value): try container.encode(value)
        case let .number(value): try container.encode(value)
        case let .bool(value): try container.encode(value)
        case .null: try container.encodeNil()
        }
    }
}

/// 一个 Native Page Stack 页面在 Telemetry 中使用的四层身份。
public struct LynxTelemetryIdentity: Codable, Equatable {
    public let navigationId: String
    public let navigationSessionId: String
    public let entryId: String
    public let pageViewId: String
    public let renderAttemptId: String
    public var activationId: String?
    public var transactionId: String?

    public init(
        navigationId: String = UUID().uuidString,
        navigationSessionId: String,
        entryId: String,
        pageViewId: String = UUID().uuidString,
        renderAttemptId: String = UUID().uuidString,
        activationId: String? = nil,
        transactionId: String? = nil
    ) {
        self.navigationId = navigationId
        self.navigationSessionId = navigationSessionId
        self.entryId = entryId
        self.pageViewId = pageViewId
        self.renderAttemptId = renderAttemptId
        self.activationId = activationId
        self.transactionId = transactionId
    }
}

/// prepare() 前冻结的候选 Bundle 快照。路径只允许在进程内使用，不能放入该模型。
public struct LynxAttemptedBundleSnapshot: Codable, Equatable {
    public let kind: String
    public let bundleSource: LynxTelemetryBundleSource
    public let lynxAppId: String?
    public let bundleName: String
    public let telemetryRouteKey: String
    public let otaOperationId: String?
    public let candidateReleaseId: String?
    public let expectedSha256: String?
    public let prepareStartedAtUnixMs: Int64
    public let attemptGeneration: Int

    public init(
        source: LynxTelemetryBundleSource,
        bundleName: String,
        telemetryRouteKey: String,
        lynxAppId: String? = nil,
        otaOperationId: String? = nil,
        candidateReleaseId: String? = nil,
        expectedSha256: String? = nil,
        prepareStartedAtUnixMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
        attemptGeneration: Int = 1
    ) {
        self.kind = "attempted"
        self.bundleSource = source
        self.lynxAppId = lynxAppId
        self.bundleName = bundleName
        self.telemetryRouteKey = telemetryRouteKey
        self.otaOperationId = otaOperationId
        self.candidateReleaseId = candidateReleaseId
        self.expectedSha256 = expectedSha256
        self.prepareStartedAtUnixMs = prepareStartedAtUnixMs
        self.attemptGeneration = attemptGeneration
    }
}

/// prepare 成功且完成 SHA/Release 选择后冻结的实际 Bundle 快照。
public struct LynxResolvedBundleSnapshot: Codable, Equatable {
    public let kind: String
    public let bundleSource: LynxTelemetryBundleSource
    public let lynxAppId: String?
    public let bundleName: String
    public let telemetryRouteKey: String
    public let releaseId: String?
    public let bundleSha256: String?
    public let engineVersion: String?
    public let localGeneration: String?
    public let rollbackFromReleaseId: String?
    public let resolvedAtUnixMs: Int64

    /// 绝对路径只在当前进程传给 LynxView，Codable 明确排除，避免日志/队列泄露路径。
    public let internalLocalPath: String?

    public init(
        source: LynxTelemetryBundleSource,
        bundleName: String,
        telemetryRouteKey: String,
        lynxAppId: String? = nil,
        releaseId: String? = nil,
        bundleSha256: String? = nil,
        engineVersion: String? = nil,
        localGeneration: String? = nil,
        rollbackFromReleaseId: String? = nil,
        resolvedAtUnixMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
        internalLocalPath: String? = nil
    ) {
        self.kind = "resolved"
        self.bundleSource = source
        self.lynxAppId = lynxAppId
        self.bundleName = bundleName
        self.telemetryRouteKey = telemetryRouteKey
        self.releaseId = releaseId
        self.bundleSha256 = bundleSha256
        self.engineVersion = engineVersion
        self.localGeneration = localGeneration
        self.rollbackFromReleaseId = rollbackFromReleaseId
        self.resolvedAtUnixMs = resolvedAtUnixMs
        self.internalLocalPath = internalLocalPath
    }

    private enum CodingKeys: String, CodingKey {
        case kind, bundleSource, lynxAppId, bundleName, telemetryRouteKey
        case releaseId, bundleSha256, engineVersion, localGeneration
        case rollbackFromReleaseId, resolvedAtUnixMs
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.kind = try container.decode(String.self, forKey: .kind)
        self.bundleSource = try container.decode(LynxTelemetryBundleSource.self, forKey: .bundleSource)
        self.lynxAppId = try container.decodeIfPresent(String.self, forKey: .lynxAppId)
        self.bundleName = try container.decode(String.self, forKey: .bundleName)
        self.telemetryRouteKey = try container.decode(String.self, forKey: .telemetryRouteKey)
        self.releaseId = try container.decodeIfPresent(String.self, forKey: .releaseId)
        self.bundleSha256 = try container.decodeIfPresent(String.self, forKey: .bundleSha256)
        self.engineVersion = try container.decodeIfPresent(String.self, forKey: .engineVersion)
        self.localGeneration = try container.decodeIfPresent(String.self, forKey: .localGeneration)
        self.rollbackFromReleaseId = try container.decodeIfPresent(String.self, forKey: .rollbackFromReleaseId)
        self.resolvedAtUnixMs = try container.decode(Int64.self, forKey: .resolvedAtUnixMs)
        // 绝对路径不会从 Wire/磁盘反序列化回来。
        self.internalLocalPath = nil
    }
}

public enum LynxTelemetryPageState: String, Codable, Equatable {
    case allocated
    case registered
    case visible
    case hidden
    case destroyed
}

public enum LynxTelemetryAppState: String, Codable, Equatable {
    case foreground
    case background
}

/// Wire 层的正交页面/App/渲染状态；页面状态不能代替 App 前后台状态。
public struct LynxTelemetryLifecycleSnapshot: Codable, Equatable {
    public let pageState: LynxTelemetryPageState
    public let appState: LynxTelemetryAppState
    public let attemptState: String
    public let activeEligible: Bool
    public let reason: String?

    public init(pageState: LynxTelemetryPageState, appState: LynxTelemetryAppState, attemptState: String, activeEligible: Bool, reason: String? = nil) {
        self.pageState = pageState
        self.appState = appState
        self.attemptState = attemptState
        self.activeEligible = activeEligible
        self.reason = reason
    }
}

public struct LynxTelemetryNavigationSnapshot: Codable, Equatable {
    public let state: String
    public let rejectionCode: String?

    public init(state: String, rejectionCode: String? = nil) {
        self.state = state
        self.rejectionCode = rejectionCode
    }
}

public struct LynxTelemetryTransitionSnapshot: Codable, Equatable {
    public let state: String
    public let terminal: LynxTelemetryTransitionTerminal?
    public let durationMs: Int64?

    public init(state: String, terminal: LynxTelemetryTransitionTerminal? = nil, durationMs: Int64? = nil) {
        self.state = state
        self.terminal = terminal
        self.durationMs = durationMs
    }
}

public struct LynxTelemetryPrivacySnapshot: Codable, Equatable {
    public let consentState: String
    public let subjectRefPresent: Bool
    public let optOut: Bool

    public init(consentState: String, subjectRefPresent: Bool, optOut: Bool) {
        self.consentState = consentState
        self.subjectRefPresent = subjectRefPresent
        self.optOut = optOut
    }
}

public enum LynxTelemetryTransitionTerminal: String, Codable, Equatable {
    case completed
    case degraded
    case cancelled
    case failed
    case notApplicable = "notApplicable"
}

/// Wire Schema 3.0.0 的本地事件对象；此阶段只交给 Noop/Debug Sink。
public struct LynxTelemetryEvent: Codable, Equatable {
    public let schemaVersion: String
    public let eventId: String
    public let eventName: String
    public let category: String
    public let source: String
    public let occurredAtUnixMs: Int64
    public let monotonicOffsetMs: Int64
    public let sequenceNo: Int
    public let navigationId: String?
    public let navigationSessionId: String?
    public let pageViewId: String?
    public let entryId: String?
    public let renderAttemptId: String?
    public let activationId: String?
    public let transactionId: String?
    public let runtimeKind: String
    public let runtimeInstanceId: String
    public let telemetryRouteKey: String
    public let hostMode: String
    public let platform: String
    public let hostApp: String
    public let appVersion: String
    public let buildNumber: String
    public let lynxSdkVersion: String
    public let engineVersion: String
    /// 兼容进程内调用方；Wire 编码只输出 lifecycle 对象，不输出这两个顶层旧字段。
    public var pageState: LynxTelemetryPageState? = nil
    public var appState: LynxTelemetryAppState? = nil
    public let lifecycle: LynxTelemetryLifecycleSnapshot?
    public let navigationAdmission: LynxTelemetryNavigationSnapshot?
    public let transition: LynxTelemetryTransitionSnapshot?
    public let privacy: LynxTelemetryPrivacySnapshot
    public let attemptedBundleSnapshot: LynxAttemptedBundleSnapshot?
    public let resolvedBundleSnapshot: LynxResolvedBundleSnapshot?
    public let sampleRate: Double
    public let samplingGroup: String
    public let samplingRuleVersion: String
    public let analysisEligible: Bool
    public let deliveryOwner: String
    public let attributes: [String: LynxTelemetryValue]

    private enum CodingKeys: String, CodingKey {
        case schemaVersion, eventId, eventName, category, source
        case occurredAtUnixMs, monotonicOffsetMs, sequenceNo
        case navigationId, navigationSessionId, pageViewId, entryId, renderAttemptId, activationId, transactionId
        case runtimeKind, runtimeInstanceId, telemetryRouteKey, hostMode, platform, hostApp, appVersion, buildNumber
        case lynxSdkVersion, engineVersion, attemptedBundleSnapshot, resolvedBundleSnapshot
        case lifecycle, navigationAdmission, transition, sampleRate, samplingGroup, samplingRuleVersion
        case analysisEligible, deliveryOwner, privacy, attributes
    }

    public init(
        eventName: String,
        category: String,
        source: String,
        occurredAtUnixMs: Int64,
        monotonicOffsetMs: Int64,
        sequenceNo: Int,
        identity: LynxTelemetryIdentity?,
        runtimeKind: String,
        runtimeInstanceId: String,
        telemetryRouteKey: String?,
        hostMode: String,
        platform: String,
        hostApp: String,
        appVersion: String,
        buildNumber: String,
        lynxSdkVersion: String,
        engineVersion: String?,
        pageState: LynxTelemetryPageState?,
        appState: LynxTelemetryAppState?,
        attemptedBundle: LynxAttemptedBundleSnapshot?,
        resolvedBundle: LynxResolvedBundleSnapshot?,
        sampleRate: Double,
        samplingGroup: String,
        samplingRuleVersion: String,
        analysisEligible: Bool = true,
        deliveryOwner: String = "internal",
        lifecycle: LynxTelemetryLifecycleSnapshot? = nil,
        navigationAdmission: LynxTelemetryNavigationSnapshot? = nil,
        transition: LynxTelemetryTransitionSnapshot? = nil,
        privacy: LynxTelemetryPrivacySnapshot = LynxTelemetryPrivacySnapshot(consentState: "unknown", subjectRefPresent: false, optOut: false),
        attributes: [String: LynxTelemetryValue]
    ) {
        self.schemaVersion = "3.0.0"
        self.eventId = "evt_" + UUID().uuidString.replacingOccurrences(of: "-", with: "")
        self.eventName = eventName
        self.category = category
        self.source = source
        self.occurredAtUnixMs = occurredAtUnixMs
        self.monotonicOffsetMs = monotonicOffsetMs
        self.sequenceNo = sequenceNo
        self.navigationId = identity?.navigationId
        self.navigationSessionId = identity?.navigationSessionId
        self.pageViewId = identity?.pageViewId
        self.entryId = identity?.entryId
        self.renderAttemptId = identity?.renderAttemptId
        self.activationId = identity?.activationId
        self.transactionId = identity?.transactionId
        self.runtimeKind = runtimeKind
        self.runtimeInstanceId = runtimeInstanceId
        self.telemetryRouteKey = telemetryRouteKey ?? "unknown"
        self.hostMode = hostMode
        self.platform = platform
        self.hostApp = hostApp
        self.appVersion = appVersion
        self.buildNumber = buildNumber
        self.lynxSdkVersion = lynxSdkVersion
        self.engineVersion = engineVersion ?? "unknown"
        self.pageState = pageState
        self.appState = appState
        self.lifecycle = lifecycle ?? pageState.map {
            LynxTelemetryLifecycleSnapshot(
                pageState: $0,
                appState: appState ?? .background,
                attemptState: resolvedBundle == nil ? "unusable" : "usable",
                activeEligible: $0 == .visible && (appState ?? .background) == .foreground && resolvedBundle != nil,
                reason: nil
            )
        }
        self.navigationAdmission = navigationAdmission
        self.transition = transition
        self.privacy = privacy
        self.attemptedBundleSnapshot = attemptedBundle
        self.resolvedBundleSnapshot = resolvedBundle
        self.sampleRate = sampleRate
        self.samplingGroup = samplingGroup
        self.samplingRuleVersion = samplingRuleVersion
        self.analysisEligible = analysisEligible
        self.deliveryOwner = deliveryOwner
        self.attributes = attributes
    }

    /// 只编码共享 Wire Schema 字段，避免把旧的顶层 pageState/appState 写成非法额外字段。
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(schemaVersion, forKey: .schemaVersion)
        try container.encode(eventId, forKey: .eventId)
        try container.encode(eventName, forKey: .eventName)
        try container.encode(category, forKey: .category)
        try container.encode(source, forKey: .source)
        try container.encode(occurredAtUnixMs, forKey: .occurredAtUnixMs)
        try container.encode(monotonicOffsetMs, forKey: .monotonicOffsetMs)
        try container.encode(sequenceNo, forKey: .sequenceNo)
        try container.encodeIfPresent(navigationId, forKey: .navigationId)
        try container.encodeIfPresent(navigationSessionId, forKey: .navigationSessionId)
        try container.encodeIfPresent(pageViewId, forKey: .pageViewId)
        try container.encodeIfPresent(entryId, forKey: .entryId)
        try container.encodeIfPresent(renderAttemptId, forKey: .renderAttemptId)
        try container.encodeIfPresent(activationId, forKey: .activationId)
        try container.encodeIfPresent(transactionId, forKey: .transactionId)
        try container.encode(runtimeKind, forKey: .runtimeKind)
        try container.encode(runtimeInstanceId, forKey: .runtimeInstanceId)
        try container.encode(telemetryRouteKey, forKey: .telemetryRouteKey)
        try container.encode(hostMode, forKey: .hostMode)
        try container.encode(platform, forKey: .platform)
        try container.encode(hostApp, forKey: .hostApp)
        try container.encode(appVersion, forKey: .appVersion)
        try container.encode(buildNumber, forKey: .buildNumber)
        try container.encode(lynxSdkVersion, forKey: .lynxSdkVersion)
        try container.encode(engineVersion, forKey: .engineVersion)
        try container.encodeIfPresent(attemptedBundleSnapshot, forKey: .attemptedBundleSnapshot)
        try container.encodeIfPresent(resolvedBundleSnapshot, forKey: .resolvedBundleSnapshot)
        try container.encodeIfPresent(lifecycle, forKey: .lifecycle)
        try container.encodeIfPresent(navigationAdmission, forKey: .navigationAdmission)
        try container.encodeIfPresent(transition, forKey: .transition)
        try container.encode(sampleRate, forKey: .sampleRate)
        try container.encode(samplingGroup, forKey: .samplingGroup)
        try container.encode(samplingRuleVersion, forKey: .samplingRuleVersion)
        try container.encode(analysisEligible, forKey: .analysisEligible)
        try container.encode(deliveryOwner, forKey: .deliveryOwner)
        try container.encode(privacy, forKey: .privacy)
        try container.encode(attributes, forKey: .attributes)
    }
}

File: android/lynx-shell/src/main/kotlin/com/example/lynxshell/telemetry/TelemetryModels.kt
package com.example.lynxshell.telemetry

import java.util.UUID

/**
 * Native 监控事件的 Wire Schema 版本。
 *
 * 这是事件格式版本，不是 Router/OTA SDK 版本。三端在 Alpha 阶段必须保持这个值一致。
 */
const val TELEMETRY_WIRE_SCHEMA_VERSION = "3.0.0"

/** Bundle 来源的低基数枚举；不能把本机绝对路径写入 Telemetry。 */
enum class TelemetryBundleSource(val wireName: String) {
    OTA("ota"),
    DIRECT_HTTPS("direct_https"),
    ASSETS("assets"),
    LOCAL_FILE("local_file"),
}

enum class TelemetryRuntimeKind(val wireName: String) {
    LYNX_VIEW("lynx_view"),
    BACKGROUND_RUNTIME("background_runtime"),
}

enum class TelemetryCategory(val wireName: String) {
    ROUTER("router"),
    NAVIGATION("navigation"),
    PAGE("page"),
    APP("app"),
    TRANSITION("transition"),
    EXPOSURE("exposure"),
    OTA("ota"),
    DIAGNOSTIC("diagnostic"),
    TELEMETRY("telemetry"),
}

enum class PageLifecycleState(val wireName: String) {
    ALLOCATED("allocated"),
    REGISTERED("registered"),
    VISIBLE("visible"),
    HIDDEN("hidden"),
    DESTROYED("destroyed"),
}

enum class AppLifecycleState(val wireName: String) {
    FOREGROUND("foreground"),
    BACKGROUND("background"),
}

enum class RenderAttemptState(val wireName: String) {
    UNUSABLE("unusable"),
    USABLE("usable"),
}

enum class TransitionTerminal(val wireName: String) {
    COMPLETED("completed"),
    DEGRADED("degraded"),
    CANCELLED("cancelled"),
    FAILED("failed"),
    NOT_APPLICABLE("notApplicable"),
}

enum class NavigationAdmission(val wireName: String) {
    REQUESTED("requested"),
    ACCEPTED("accepted"),
    REJECTED("rejected"),
}

/**
 * 页面访问的 Native 身份。只有 Native 生成这些 ID；Lynx 页面提交同名字段时会被忽略。
 * pageViewId 可以为空，适用于尚未创建页面实例的 navigation_rejected。
 */
data class TelemetryIdentity(
    val navigationId: String = newTelemetryId(),
    val navigationSessionId: String? = null,
    val entryId: String? = null,
    val pageViewId: String? = null,
    val renderAttemptId: String? = null,
    val activationId: String? = null,
    val transactionId: String? = null,
) {
    companion object {
        fun forPage(
            entryId: String,
            navigationId: String = newTelemetryId(),
            navigationSessionId: String? = null,
            pageViewId: String = newTelemetryId(),
            transactionId: String? = null,
        ): TelemetryIdentity = TelemetryIdentity(
            navigationId = navigationId,
            navigationSessionId = navigationSessionId,
            entryId = entryId,
            pageViewId = pageViewId,
            transactionId = transactionId,
        )
    }
}

/** prepare() 前冻结的候选 Bundle 信息，用于解释缺包、下载、校验和取消。 */
data class AttemptedBundleSnapshot(
    val bundleSource: TelemetryBundleSource,
    val lynxAppId: String? = null,
    val bundleName: String,
    val telemetryRouteKey: String,
    val otaOperationId: String? = null,
    val candidateReleaseId: String? = null,
    val expectedSha256: String? = null,
    val prepareStartedAtUnixMs: Long,
    val attemptGeneration: Long,
)

/** prepare 成功且本地 SHA 校验完成后冻结的 Bundle 信息。 */
data class ResolvedBundleSnapshot(
    val bundleSource: TelemetryBundleSource,
    val lynxAppId: String? = null,
    val bundleName: String,
    val telemetryRouteKey: String,
    val releaseId: String? = null,
    val bundleSha256: String? = null,
    val engineVersion: String? = null,
    val localGeneration: String? = null,
    val rollbackFromReleaseId: String? = null,
    val resolvedAtUnixMs: Long,
    /** 仅进程内加载使用，永远不会出现在 TelemetryEvent 或 Debug 输出。 */
    val internalLocalPath: String? = null,
)

/** 仅允许安全的低基数字段进入 Event；URL、绝对路径和原始错误不会从这里导出。 */
data class TelemetryBundleSnapshotWire(
    val kind: String,
    val bundleSource: String,
    val lynxAppId: String?,
    val bundleName: String,
    val telemetryRouteKey: String,
    val otaOperationId: String?,
    val releaseId: String?,
    val sha256: String?,
    val prepareStartedAtUnixMs: Long? = null,
    val attemptGeneration: Long? = null,
    val engineVersion: String? = null,
    val localGeneration: String? = null,
    val rollbackFromReleaseId: String? = null,
    val resolvedAtUnixMs: Long? = null,
) {
    /** 旧平台适配器使用 source 命名时仍能读取统一字段。 */
    val source: String
        get() = bundleSource
)

fun AttemptedBundleSnapshot.toWireSnapshot(): TelemetryBundleSnapshotWire =
    TelemetryBundleSnapshotWire(
        kind = "attempted",
        bundleSource = bundleSource.wireName,
        lynxAppId = lynxAppId,
        bundleName = bundleName,
        telemetryRouteKey = telemetryRouteKey,
        otaOperationId = otaOperationId,
        releaseId = candidateReleaseId,
        sha256 = expectedSha256,
        prepareStartedAtUnixMs = prepareStartedAtUnixMs,
        attemptGeneration = attemptGeneration,
    )

fun ResolvedBundleSnapshot.toWireSnapshot(): TelemetryBundleSnapshotWire =
    TelemetryBundleSnapshotWire(
        kind = "resolved",
        bundleSource = bundleSource.wireName,
        lynxAppId = lynxAppId,
        bundleName = bundleName,
        telemetryRouteKey = telemetryRouteKey,
        otaOperationId = null,
        releaseId = releaseId,
        sha256 = bundleSha256,
        engineVersion = engineVersion,
        localGeneration = localGeneration,
        rollbackFromReleaseId = rollbackFromReleaseId,
        resolvedAtUnixMs = resolvedAtUnixMs,
    )

/** 采样信息必须随事件携带；同一个 pageView 的页面、曝光和点击共用一个 cohort。 */
data class SamplingDecision(
    val sampleRate: Double,
    val samplingGroup: String,
    val samplingRuleVersion: String,
    val sampled: Boolean,
    val inclusionProbability: Double,
)

/** 宿主侧的低基数上下文。不得放入账号、URL query、原始 params 或绝对路径。 */
data class TelemetryHostContext(
    val runtimeKind: TelemetryRuntimeKind = TelemetryRuntimeKind.LYNX_VIEW,
    val runtimeInstanceId: String = newTelemetryId(),
    val telemetryRouteKey: String = "unknown",
    val hostMode: String = "activity",
    val platform: String = "android",
    val hostApp: String = "unknown",
    val appVersion: String = "unknown",
    val buildNumber: String = "unknown",
    val lynxSdkVersion: String = "unknown",
    val engineVersion: String = "unknown",
)

data class TelemetryLifecycleSnapshot(
    val pageState: PageLifecycleState,
    val appState: AppLifecycleState,
    val attemptState: RenderAttemptState,
    val activeEligible: Boolean,
    val reason: String? = null,
)

data class TelemetryNavigationSnapshot(
    val state: NavigationAdmission,
    val rejectionCode: String? = null,
)

data class TelemetryTransitionSnapshot(
    val state: String = "waitingTarget",
    val terminal: TransitionTerminal? = null,
    val durationMs: Long? = null,
)

data class TelemetryPrivacySnapshot(
    val consentState: String = "unknown",
    val subjectRefPresent: Boolean = false,
    val optOut: Boolean = false,
)

/**
 * 旁路事件模型。toWireMap() 只导出 Wire Schema 允许字段，尤其不会导出 Resolved Snapshot 的本地路径。
 */
data class TelemetryEvent(
    val eventId: String = newTelemetryEventId(),
    val eventName: String,
    val category: TelemetryCategory,
    val source: String = "native",
    val occurredAtUnixMs: Long,
    val monotonicOffsetMs: Long,
    val sequenceNo: Long,
    val identity: TelemetryIdentity,
    val host: TelemetryHostContext,
    val attemptedBundle: TelemetryBundleSnapshotWire? = null,
    val resolvedBundle: TelemetryBundleSnapshotWire? = null,
    val sampling: SamplingDecision,
    val lifecycle: TelemetryLifecycleSnapshot? = null,
    val navigationAdmission: TelemetryNavigationSnapshot? = null,
    val transition: TelemetryTransitionSnapshot? = null,
    val privacy: TelemetryPrivacySnapshot = TelemetryPrivacySnapshot(),
    val analysisEligible: Boolean = true,
    val samplingReason: String = if (sampling.sampled) "stable_cohort" else "not_sampled",
    val deliveryOwner: String = "internal",
    val otaOperationId: String? = null,
    val runtimeGeneration: Long? = null,
    val attributes: Map<String, String> = emptyMap(),
) {
    /** 给 Debug Sink 和后续上传适配器使用的平面 Wire 结构。 */
    fun toWireMap(): Map<String, Any?> {
        val wire = linkedMapOf<String, Any?>(
        "schemaVersion" to TELEMETRY_WIRE_SCHEMA_VERSION,
        "eventId" to eventId,
        "eventName" to eventName,
        "category" to category.wireName,
        "source" to source,
        "occurredAtUnixMs" to occurredAtUnixMs,
        "monotonicOffsetMs" to monotonicOffsetMs,
        "sequenceNo" to sequenceNo,
        "navigationId" to identity.navigationId,
        "navigationSessionId" to identity.navigationSessionId,
        "pageViewId" to identity.pageViewId,
        "entryId" to identity.entryId,
        "renderAttemptId" to identity.renderAttemptId,
        "activationId" to identity.activationId,
        "transactionId" to identity.transactionId,
        "runtimeKind" to host.runtimeKind.wireName,
        "runtimeInstanceId" to host.runtimeInstanceId,
        "telemetryRouteKey" to host.telemetryRouteKey,
        "hostMode" to host.hostMode,
        "platform" to host.platform,
        "hostApp" to host.hostApp,
        "appVersion" to host.appVersion,
        "buildNumber" to host.buildNumber,
        "lynxSdkVersion" to host.lynxSdkVersion,
        "engineVersion" to host.engineVersion,
        "runtimeGeneration" to runtimeGeneration,
        "otaOperationId" to otaOperationId,
        "attemptedBundleSnapshot" to attemptedBundle?.toWireMap(),
        "resolvedBundleSnapshot" to resolvedBundle?.toWireMap(),
        "lifecycle" to lifecycle?.toWireMap(),
        "navigationAdmission" to navigationAdmission?.toWireMap(),
        "transition" to transition?.toWireMap(),
        "sampleRate" to sampling.sampleRate,
        "samplingGroup" to sampling.samplingGroup,
        "samplingRuleVersion" to sampling.samplingRuleVersion,
        "inclusionProbability" to sampling.inclusionProbability,
        "analysisEligible" to analysisEligible,
        "samplingReason" to samplingReason,
        "deliveryOwner" to deliveryOwner,
        "privacy" to privacy.toWireMap(),
            "attributes" to attributes,
        )
        // JSON Schema 的可选对象不能以 null 形式发送；省略而不是写入 null，避免服务端拒绝整条事件。
        return wire.filterValues { it != null }
    }
}

private fun TelemetryBundleSnapshotWire.toWireMap(): Map<String, Any?> {
    val common = linkedMapOf<String, Any?>(
        "kind" to kind,
        "bundleSource" to bundleSource,
        "lynxAppId" to lynxAppId,
        "bundleName" to bundleName,
        "telemetryRouteKey" to telemetryRouteKey,
    )
    if (kind == "attempted") {
        common["otaOperationId"] = otaOperationId
        common["candidateReleaseId"] = releaseId
        common["expectedSha256"] = sha256
        common["prepareStartedAtUnixMs"] = prepareStartedAtUnixMs
        common["attemptGeneration"] = attemptGeneration
    } else {
        common["releaseId"] = releaseId
        common["bundleSha256"] = sha256
        common["engineVersion"] = engineVersion
        common["localGeneration"] = localGeneration
        common["rollbackFromReleaseId"] = rollbackFromReleaseId
        common["resolvedAtUnixMs"] = resolvedAtUnixMs
    }
    return common
}

private fun TelemetryLifecycleSnapshot.toWireMap(): Map<String, Any?> = linkedMapOf(
    "pageState" to pageState.wireName,
    "appState" to appState.wireName,
    "attemptState" to attemptState.wireName,
    "activeEligible" to activeEligible,
    "reason" to reason,
)

private fun TelemetryNavigationSnapshot.toWireMap(): Map<String, Any?> = linkedMapOf(
    "state" to state.wireName,
    "rejectionCode" to rejectionCode,
)

private fun TelemetryTransitionSnapshot.toWireMap(): Map<String, Any?> = linkedMapOf(
    "state" to state,
    "terminal" to terminal?.wireName,
    "durationMs" to durationMs,
)

private fun TelemetryPrivacySnapshot.toWireMap(): Map<String, Any?> = linkedMapOf(
    "consentState" to consentState,
    "subjectRefPresent" to subjectRefPresent,
    "optOut" to optOut,
)

/** 确定性 ID 生成器，后续可以在单测中注入固定实现。 */
fun newTelemetryId(): String = UUID.randomUUID().toString()

/** Wire Schema 要求事件 ID 具备 evt_ 前缀；页面身份仍使用普通 Native ID。 */
fun newTelemetryEventId(): String = "evt_${newTelemetryId()}"
